app/src/main/java/com/eveningoutpost/dexdrip/NFCReaderX.java

```diff
@@ -277,125 +408,17 @@ protected Tag doInBackground(Tag... params) {
-                                    cmd = new byte[]{0x02, 0x23, (byte) i, 0x02};
-                                }
-
-                                byte[] replyBlock;
-                                Long time = System.currentTimeMillis();
-                                while (true) {
-                                    try {
-                                        replyBlock = nfcvTag.transceive(cmd);
-                                        break;
-                                    } catch (IOException e) {
-                                        if ((System.currentTimeMillis() > time + 2000)) {
-                                            Log.e(TAG, "tag read timeout");
-                                            JoH.static_toast_short("NFC read timeout");
-                                            vibrate(context, 3);
-                                            return null;
-                                        }
-                                        Thread.sleep(100);
-                                    }
-                                }
-
-                                if (d)
-                                    Log.d(TAG, "Received multiblock reply, offset: " + i + " sized: " + replyBlock.length);
-                                if (d)
-                                    Log.d(TAG, HexDump.dumpHexString(replyBlock, 0, replyBlock.length));
-                                if (replyBlock.length != correct_reply_size) {
-                                    Log.e(TAG, "Incorrect block size: " + replyBlock.length + " vs " + correct_reply_size);
-                                    JoH.static_toast_short("NFC invalid data - try again");
-                                    if (!addressed) {
-                                        if (PersistentStore.incrementLong("nfc-address-failures") > 2) {
-                                            Home.setPreferencesBoolean("use_nfc_any_tag", false);
-                                            JoH.static_toast_short("Turned off any-tag feature");
-                                        }
-                                    }
-                                    vibrate(context, 3);
-                                    return null;
-                                }
-                                if (addressed) {
-                                    for (int j = 0; j < 3; j++) {
-                                        System.arraycopy(replyBlock, 2 + (j * 9), data, i * 8 + (j * 8), 8);
-                                    }
-                                } else {
-                                    System.arraycopy(replyBlock, 1, data, i * 8, replyBlock.length - 1);
-                                }
-                            }
+                        boolean readPatchSucceeded;
+                        if (Home.getPreferencesBooleanDefaultFalse("external_blukon_algorithm")) {
+                        	readPatchSucceeded = LibreOOPAlgorithm.readOOPData(nfcvTag);
```

Comment:
Instead why not change the reading block count from 40 to 42 if `external_blukon_algorithm` is set and use the local `readPatch` and then call `SendData` directly to eliminate duplicate code in the OOP class? Possibly do an array copy to get the buffer the exact size the interface is expecting. Maybe this is already what you are planning to do..

Reply:
Well, read the code of nfcreaderx.readPatch again.
I guess that if all the conditions below will be true: 
1) nfc_test_diagnostic is off  
2) multiblock is on.
3) addressed is off
4) Increase the number of blocks that is read from 42 to 43 (that does not divide in 3)
Than the code should do the exact same thing.

So I'll change the function if that is what you think that I should do, but do you think that it will make the code more readable.

Reply:
As far as I can see the code in `LibreOOPAlgorithm.readPatch` duplicates the nfc transmission functionality of the existing `readPatch` method which has greater flexibility which was put in for speed and compatibility reasons. 

The only real difference I can see is that `LibreOOPAlgorithm.readPatch` reads slightly more data. My thinking is that this duplicate code should be removed and I assume that the existing `readPatch` would want to be modified slightly to read the additional data blocks as I assume your external interface needs *exactly* that many bytes of data, which is what I was getting at earlier.

If multiples of 3 is a problem then just read an extra block and truncate the resulting byte array before returning it or we can do something to read one or two singles blocks at the end if in multi-block mode if it is absolutely necessary.

If you *need* an initial preparation command like is in `getPatchInfo` then you can just inject the byte sequence in to the existing `diag` byte array and it will be transmitted before the main read occurs. As far as I am aware this is not required for successful read operations, but if you have tested using this already and feel it wont work without it then that is where I would suggest utilizing that functionality. 

Ideally have a method in `LibreOOPAlgorithm` which returns the byte sequence and populate `diag` from that if the preference item is set and then pass the diag `replyBlock` back along with the main data block or if you need to cancel before that then make a call to `resultOk` and skip/return early based on a failure there.

The idea is to keep the nfc reading in one place and any specific data processing in another class as much as possible.